File: androidApp/src/main/java/com/luisma/conexiones/android/features/levels/LevelsViewModel.kt
package com.luisma.conexiones.android.features.levels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.viewModelFactory
import com.luisma.conexiones.android.router.RoutePayload
import com.luisma.conexiones.android.router.RouterService
import com.luisma.conexiones.android.router.Routes
import com.luisma.conexiones.models.BasicScreenState
import com.luisma.conexiones.services.IUserProfileService
import com.luisma.conexiones.services.PaginationService
import com.luisma.conexiones.services.game.GamesLevelsService
import com.luisma.conexiones.services.gamesLevelsService
import com.luisma.conexiones.services.paginationService
import com.luisma.conexiones.services.userProfileService
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class LevelsViewModel(
    private val gamesLevelsService: GamesLevelsService,
    private val paginationService: PaginationService,
    private val userProfileService: IUserProfileService,
    private val routerService: RouterService
) : ViewModel() {

    companion object {
        val Factory: ViewModelProvider.Factory = viewModelFactory {
            addInitializer(LevelsViewModel::class) {
                LevelsViewModel(
                    gamesLevelsService = gamesLevelsService(),
                    paginationService = paginationService(),
                    userProfileService = userProfileService(),
                    routerService = RouterService
                )
            }
        }
    }

    private val _state = MutableStateFlow(LevelsState.initial())
    val state = _state.asStateFlow()

    init {
        initVM()
    }

    fun sendEvent(event: LevelsEvents) {
        when (event) {
            LevelsEvents.SetInitialScrollDone -> setInitialScrollDone()
            LevelsEvents.RefreshScreen -> refreshScreen()
            LevelsEvents.GetPreviousPage -> getPreviousPage()
            LevelsEvents.GetNextPage -> getNextPage()
            LevelsEvents.DismissTutorial -> dismissTutorial()
            is LevelsEvents.RefreshFromFlag -> refreshFromFlag(flag = event.flag)
            is LevelsEvents.OnVisibilityChangePlayingCard -> onVisibilityChange(show = event.show)
            is LevelsEvents.OnTapLevel -> onTapLevel(gameId = event.gameId)
        }
    }

    private fun initVM() {
        _state.update {
            LevelsState.initial().copy(
                refreshOnBackControlFlag = _state.value.refreshOnBackControlFlag,
            )
        }
        viewModelScope.launch {
            val responsePlayingDef = async { gamesLevelsService.getLevelsPlayingPage() }
            val responseLivesDef = async { userProfileService.getLives() }
            val appWasOpenedBeforeDef = async { userProfileService.appWasOpenedBefore() }
            val responsePlaying = responsePlayingDef.await()
            val responseLives = responseLivesDef.await()
            val appWasOpenedBefore = appWasOpenedBeforeDef.await()

            _state.update {
                it.copy(
                    screenState = BasicScreenState.Success,
                    games = responsePlaying.games,
                    startPage = responsePlaying.page,
                    initialScrollIdx = responsePlaying.firstPagePlayingLevelIdxOffset,
                    reachFirstPage = paginationService.reachFirstPageByPageState(
                        responsePlaying.pageState
                    ),
                    reachLastPage = paginationService.reachLastPageByPageState(
                        responsePlaying.pageState
                    ),
                    fetchedPages = setOf(responsePlaying.page),
                    playingRowId = responsePlaying.playingRowId,
                    playingRowIdx = gamesLevelsService.getIndexOfRowId(
                        games = responsePlaying.games,
                        rowId = responsePlaying.playingRowId
                    ),
                    playingRowIsShowing = true,
                    lives = responseLives,
                    openTutorial = !appWasOpenedBefore,
                )
            }
        }
    }

    private fun refreshScreen() {
        initVM()
    }

    private fun dismissTutorial() {
        if (_state.value.openTutorial) {
            _state.update {
                it.copy(openTutorial = false)
            }
        }
    }

    private fun setInitialScrollDone() {
        if (!_state.value.initialScrollDone) {
            _state.update {
                it.copy(
                    initialScrollDone = true,
                )
            }
        }
    }

    private fun getPreviousPage() {
        _state.update {
            it.copy(
                screenState = BasicScreenState.Loading,
            )
        }

        viewModelScope.launch {
            val fetchedPages = paginationService.getPreviousPage(
                fetchedPages = _state.value.fetchedPages
            )
            val responsePlaying = gamesLevelsService.getLevelsByPage(
                page = fetchedPages.first()
            )
            val games = responsePlaying.games + _state.value.games
            _state.update {
                it.copy(
                    screenState = BasicScreenState.Success,
                    games = games,
                    reachFirstPage = paginationService.reachFirstPageByPageState(
                        responsePlaying.pageState
                    ),
                    fetchedPages = fetchedPages,
                    playingRowIdx = gamesLevelsService.getIndexOfRowId(
                        games = games,
                        rowId = _state.value.playingRowId
                    ),
                    playingRowIsShowing = false
                )
            }
        }
    }

    private fun getNextPage() {
        viewModelScope.launch {
            val fetchedPages = paginationService.getNextPage(
                fetchedPages = _state.value.fetchedPages
            )
            val responsePlaying = gamesLevelsService.getLevelsByPage(
                page = fetchedPages.last()
            )
            val games = _state.value.games + responsePlaying.games
            _state.update {
                it.copy(
                    games = games,
                    reachLastPage = paginationService.reachLastPageByPageState(
                        responsePlaying.pageState
                    ),
                    fetchedPages = fetchedPages,
                    playingRowIdx = gamesLevelsService.getIndexOfRowId(
                        games = games,
                        rowId = _state.value.playingRowId
                    ),
                )
            }
        }

    }

    private fun onVisibilityChange(show: Boolean) {
        if (show == _state.value.playingRowIsShowing) {
            return
        }
        _state.update {
            it.copy(playingRowIsShowing = show)
        }
    }

    private fun onTapLevel(gameId: Int) {
        val canGo = gamesLevelsService.canGoToLevel(
            gameId = gameId,
            games = _state.value.games,
            lives = _state.value.lives,
        )
        if (!canGo) {
            return
        }

        viewModelScope.launch {
            routerService.setRoute(
                RoutePayload(
                    route = Routes.Game,
                    payload = gameId.toString()
                )
            )
        }
    }

    private fun refreshFromFlag(flag: String) {
        if (_state.value.refreshOnBackControlFlag == flag) {
            return
        }

        _state.update {
            it.copy(refreshOnBackControlFlag = flag)
        }

        refreshScreen()
    }

}

File: iosApp/iosApp/Features/Levels/LevelsViewModel.swift
//
//  LevelsViewModel.swift
//  iosApp
//
//  Created by Luis Manuel García Fernández on 22/8/24.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import shared

class LevelsViewModel: ObservableObject {
    @Published public private(set) var state: LevelsState
    
    private let gameLevelsService: GamesLevelsService
    private let userProfileService: IUserProfileService
    private let paginationService: PaginationService
    private let router = Router.instance
    
    init(
        state: LevelsState = .init(),
        gameLevelsService: GamesLevelsService = ServicesLocatorKt.gamesLevelsService(),
        userProfileService: IUserProfileService = ServicesLocatorKt.userProfileService(),
        paginationService: PaginationService = ServicesLocatorKt.paginationService()
    ) {
        self.state = state
        self.gameLevelsService = gameLevelsService
        self.userProfileService = userProfileService
        self.paginationService = paginationService
    }
    
    func sendEvent(event: LevelsEvents) {
        switch event {
        case .onAppear:
            initViewModel(resetState: false)
        case .refreshScreen:
            initViewModel(resetState: true)
        case .initialScrollDone:
            initialScrollDone()
        case .previousPage:
            previousPage()
        case .nextPage:
            nextPage()
        case .onVisibilityChangePlayingCard(show: let show):
            onVisibilityChangePlayingCard(show: show)
        case .goToLevel(gameId: let gameId):
            goToLevel(gameId: gameId)
        }
    }
    
    private func onVisibilityChangePlayingCard(show: Bool) {
        if state.playingRowIsShowing == show {
            return
        }
        
        Task {
            await MainActor.run {
                state.playingRowIsShowing = show
            }
        }
    }
    
    private func initViewModel(resetState: Bool) {
        Task {
            if resetState {
                await MainActor.run {
                    state = .init()
                }
            }
            
            async let playingPageResp = gameLevelsService.getLevelsPlayingPage()
            async let livesResp = userProfileService.getLives()
            async let appWasOpenedResp = userProfileService.appWasOpenedBefore()
            
            let playingPage = try await (playingPageResp)
            let lives = try await (livesResp)
            let appWasOpened = try await (appWasOpenedResp)
            let playingRowId = playingPage.playingRowId
            
            await MainActor.run {
                state.games = playingPage.games
                state.startPage = Int(playingPage.page)
                state.initialScrollIdx = Int(playingPage.firstPagePlayingLevelIdxOffset
                )
                state.reachFirstPage = paginationService.reachFirstPageByPageState(
                    pageState: playingPage.pageState
                )
                state.reachLastPage = paginationService.reachLastPageByPageState(
                    pageState: playingPage.pageState
                )
                state.fetchedPages = [Int(playingPage.page)]
                state.playingRowId = Int(playingRowId)
                state.playingRowIdx = Int(gameLevelsService.getIndexOfRowId(
                    games: playingPage.games,
                    rowId: Int32(playingRowId)
                ))
                state.playingRowIsShowing = true
                state.lives = Int(truncating: lives)
                state.openTutorial = !Bool(truncating: appWasOpened)
                state.screenState = .success
            }
        }
    }
    
    private func initialScrollDone() {
        if state.initialScrollDone {
            return
        }
        
        state.initialScrollDone = true
    }
    
    private func previousPage() {
        Task {
            await MainActor.run {
                state.screenState = .loading
            }

            let fetchedPages = paginationService.getPreviousPage(
                fetchedPages: Set(state.fetchedPages.map { value in KotlinInt(value: Int32(value)) })
            )
            let fetchedPagesSorted = fetchedPages.map {
                value in Int(truncating: value)
            }.sorted(by: <)
            let responsePlaying = try await gameLevelsService.getLevelsByPage(
                page: Int32(fetchedPagesSorted.first ?? 0)
            )

            await MainActor.run {
                state.games = responsePlaying.games + state.games
                state.reachFirstPage = paginationService.reachFirstPageByPageState(
                    pageState: responsePlaying.pageState
                )
                state.fetchedPages = Set(fetchedPagesSorted)
                state.playingRowIdx = Int(gameLevelsService.getIndexOfRowId(
                    games: state.games,
                    rowId: Int32(state.playingRowId)
                ))
                state.playingRowIsShowing = false
                state.screenState = .success
            }
        }
    }
    
    private func nextPage() {
        Task {
            let fetchedPages = paginationService.getNextPage(
                fetchedPages: Set(state.fetchedPages.map { value in KotlinInt(value: Int32(value)) })
            )
            let fetchedPagesSorted = fetchedPages.map {
                value in Int(truncating: value)
            }.sorted(by: <)
            
            let resposePlaying = try await gameLevelsService.getLevelsByPage(
                page: Int32(fetchedPagesSorted.last ?? 0)
            )
            await MainActor.run {
                state.games = state.games + resposePlaying.games
                state.reachLastPage = paginationService.reachLastPageByPageState(
                    pageState: resposePlaying.pageState
                )
                state.fetchedPages = Set(fetchedPagesSorted)
                state.playingRowIdx = Int(gameLevelsService.getIndexOfRowId(
                    games: state.games,
                    rowId: Int32(state.playingRowId)
                ))
            }
        }
    }
    
    private func goToLevel(gameId: Int) {
        Task {
            let canGo = gameLevelsService.canGoToLevel(
                games: state.games,
                lives: Int32(state.lives),
                gameId: Int32(gameId)
            )
            if !canGo {
                return
            }
            await router.goTo(page: .Game(gameId: gameId))
        }
    }
}
